torch/csrc/autograd/function.h

```diff
@@ -23,164 +17,360 @@
 using IndexRange = std::pair<size_t, size_t>;
 
-namespace detail {
-struct MakeNextFunctionList : IterArgs<MakeNextFunctionList> {
-  function_list next_functions;
-  using IterArgs<MakeNextFunctionList>::operator();
-  void operator()(const Variable& variable) {
-    if (variable.defined()) {
-      next_functions.push_back(variable.gradient_edge());
-    } else {
-      next_functions.emplace_back();
-    }
-  }
-};
-} // namespace detail
-
-// Returns true if any of the variables in the list require a gradient.
-inline bool any_variable_requires_grad(const variable_list& variables) {
-  return std::any_of(
-      variables.begin(), variables.end(), [](const Variable& variable) {
-        return variable.requires_grad();
-      });
-}
-
-template <typename... Variables>
-function_list get_next_functions(Variables&&... variables) {
-  if (!GradMode::is_enabled()) return {};
-  detail::MakeNextFunctionList make;
-  make.apply(std::forward<Variables>(variables)...);
-  return std::move(make.next_functions);
-}
+///~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+///                               Function
+///~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// A `Function` is an abstract class that represents an operation taking zero
+/// or more input `Variable`s and producing zero or more output `Variable`s. All
+/// functions in PyTorch's autograd machinery derive from this class and
+/// override its `apply` method. Instances of such subclasses will then be
+/// invokeable via the call operator.
+///
+///                    Functions in the Autograd Graph
+///~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// When viewing the autograd system as a graph, `Function`s are the vertices or
+/// nodes, connected to each other via (directed) `Edge`s, which themselves are
+/// represented via (`Function`, input_nr) pairs. `Variable`s are the outputs to
+/// and inputs of `Function`s, and travel between these edges during execution
+/// of the graph. When two or more `Edge`s (from different sources) point at the
+/// same input to a `Function`, the values produced along all of these edges are
+/// implicitly summed prior to being forwarded to the target `Function`.
+///
+///                              Hierarchy
+///~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// Subclasses will often represent forward passes (functions) or backward
+/// passes (derivatives). Note, however, that due to the very general
+/// definition of a `Function` taking *zero* or more inputs and producing
+/// *zero* or more outputs, uses of `Function`s are flexible and extend beyond
+/// purely mathematical operations. For example, the `AccumulateGrad` function
+/// is a *sink*: it takes one input, but produces no outputs, instead adding
+/// the input to its internal accumulator as a side effect. At the other
```

Comment:
nit: there's no *internal accumulator* in AccumulateGrad. It takes it's `var` reference, traverses to its `.grad` attr, and adds it there.